Add a history modal for browsing/viewing notebook history

wlach opened the pull request.

wlach commented:
Note: this interface doesn't show autosaved changes in the browser that have not yet been sync'ed to the server. It probably should...

bcolloran commented on src/components/modals/history-modal.jsx:
```diff
@@ -0,0 +1,189 @@
+                onClick={() => this.props.revisionClicked(revision.id)}
+                selected={this.props.selectedRevisionId === revision.id}
+              >
+                {format(new Date(revision.created), "MMM dd, uuuu HH:MM")}
```
:+1:

bcolloran commented on src/state-schemas/state-schema.js:
```diff
@@ -111,6 +111,10 @@ export const stateProperties = {
     type: "boolean",
     default: false
   },
+  gettingNotebookHistory: {
```
i see where this bit of state is set, but i'm not spotting where it is read/used?

bcolloran commented on src/reducers/notebook-reducer.js:
```diff
@@ -103,6 +103,28 @@ const notebookReducer = (state = newNotebook(), action) => {
+      return Object.assign({}, state, {
+        gettingNotebookHistory: false,
+        notebookRevisionList: undefined,
+        errorGettingNotebookHistory: true
```
it looks like maybe this bit of state is not used anywhere, and it's not in the state schema -- maybe it's left over from something you tried out earlier?

wlach replied:
this was half-completed code for error handling. now done.

bcolloran commented:
Looks good for the MVP.  I have a couple suggestions re the code itself.

+1 on hamilton's comments, i think those are good UX directions to follow. I think that most of them can be followed up in subsequent PRs after we land this, but I agree that it would be nice to do a tiny bit of UI work to get this fit in a little better with the rest of the NB before we merge. The things that really jump out at me are:
(1) the blue list highlights
(2) the extra padding around the list and diff viewer (these should run to the edge of the modal, consistent with the rest of the "material UI" look we have)
(3) +1 on hamilton's suggestion of a "revisions" header (using our beautiful gradient!)

![image](https://user-images.githubusercontent.com/2380975/52678460-980f3500-2ee6-11e9-8907-f2150306e587.png)


I'll add one other non-blocking thing for future consideration: it actually took me a long time to realize that i was looking at the diff vs the version prior to the one selected rather than the diff between the selected version and the current version. we may want to consider somehow indicating both version that are being considered (i suspect eventually we'll want to be able to diff between arbitrary versions)

hamilton commented:
To @bcolloran's point about the confusion: 

From a UX perspective, it would be great to have above the diff itself something that says `comparing {some string that specifies version, maybe date-time} to current` or something like that so that it is clear what's being observed. Thinking of something similar to to what github does w/ a PR - "merging my-branch into master" - very useful bit of information that shows source and target.

wlach commented:
Ok, I think this is now ready for a proper review. I refactored things as you suggested, added error handling, and some of the stylistic changes that @hamilton suggested. I also added *some* tests, for some of the more complicated pieces. I could add more, but I am getting a feeling that I could spend a *very* long time testing each component and interaction, for little practical benefit. I think we should probably set some sort of reasonable limit here.

Here's what this component looks like now:

![image](https://user-images.githubusercontent.com/20569/53130962-90304000-3539-11e9-9174-6a3f9b185e06.png)

wlach commented:
Maybe more tests + stylistic changes could be done in a followup?

bcolloran commented on src/components/modals/revision-list.jsx:
```diff
@@ -0,0 +1,58 @@
+              </ListItem>
+            ))}
+        </List>
+      </RevisionListContainer>
```
it would be good practice to write some test for this. not a ton, but at least a couple to verify that that correct items are selected and that the correct number of items are present. console-pane.test.js might be a place to look.

i also think that this element should probably be connected directly to the store. none of the things that are passed down here as props from HistoryModalUnconnected actually need to be passed in at that point (that component just passes them straight through via it's mapStateToProps, which is a good sign that a child component should connect to the store). also, the onclick function used here can be a method of this component, since the selection of items is really a concern of the list, not the modal.

refactoring that stuff should slightly streamline the history modal and separate these concerns a bit better. it will also give you two smaller and simpler mapStateToProps functions to test

wlach replied:
Hmm, I don't agree with connecting this to the store, I think it's much easier to reason about this component as a display-only thing. I think it would be confusing to have three different connections to the redux store for what is essentially one component.

bcolloran replied:
haha, i disagree with your disagreement :-)

we will probably have to rewrite this component substantially a few times, and as i mention, rewrites are much easier if more things derive their display state from store. this is the convention we've adopted throughout the codebase, and it's really quite nice. I'd ask that you trust me on this one and give it a shot :-)

bcolloran commented on src/components/modals/revision-diff.jsx:
```diff
@@ -0,0 +1,41 @@
+
+  render() {
+    if (
+      this.props.gettingRevisionContent ||
```
this component should be connected to the store directly. as with the history list, the store knows everything that this component needs to render, and by connecting it directly we'll end up with more simple mapStateToProps functions, which is better than fewer complex ones.

also, i think that `gettingRevisionContent` is not actually needed. If the only states we choose to display i this component are (a) gettin revisions (b) no diff (c) some diff, then we can deduce that we are in the getting revisions state if either  `prevRev===undef || currRev===undef`. so I think we can remove all references to this variable.

(note tho that if you wish to admit case  (d) error getting revisions, then it might make sense to look at a `revisionContentFetchStatus` variable, as i suggest elsewhere)

wlach replied:
Again, I don't think connecting this to the redux store would improve things. You may be right that we could infer the fetching status in some other way though, i'll look into it.

bcolloran replied:
again, connecting components to the store is the convention we've adopted, and our experience has shown that it has many advantages. please give it a shot :-)

wlach commented on src/actions/actions.js:
```diff
@@ -326,6 +331,65 @@ export function saveNotebookToServer() {
   };
 }
 
+function getRequiredRevisionContent(state, dispatch) {
```
It's not a separate action, but a helper function called by two actions. I'm not really sure what the purpose of calling dispatch on it is? I could prefix the function with an underscore if it helps make that more clear, though we don't do that elsewhere in the codebase.

bcolloran replied:
hmm, i guess it's kind of a semantic distinction -- it's a function that results in updates to the redux store by calling dispatch and that has a reference to the current state, which is basically what a thunk does. but i guess the difference here is that it's not meant to be called anywhere outside of this helper function. ok, i buy that, sounds good! :-)

wlach commented on src/state-schemas/state-schema.js:
```diff
@@ -182,6 +186,38 @@ export const stateProperties = {
+      selectedRevision: {
+        type: "number"
+      },
+      revisionContent: {
```
I'm seperating out the fetching of the revision metadata + their content because the sum of every revision + its content might be prohibitively large in certain cases. I can do what you suggest though wrt additionalProperties.

bcolloran replied:
right, but you can start with the `content: undefined`, and then lazily add it when you fetch the revision -- it's just a matter of where you put the data. putting the actual content within the rest of the revisionList seems cleaner overall.

wlach replied:
Storing the content within a separate object simplifies look-up when it comes time to actually retrieve the content for display. It's also easier to just blow away the revisionList structure when we want to download/refresh a new one, since nothing major will be lost. I'd rather leave this part as-is.

bcolloran replied:
gotcha, sounds good :+1:

bcolloran replied:
oops, i think the comment about adding `additionalProperties: {type: "string"}` to the schema for `revisionContent` got lost in the thread

wlach commented on src/state-schemas/state-schema.js:
```diff
@@ -182,6 +186,38 @@ export const stateProperties = {
+          properties: {
+            id: { type: "integer" },
+            title: { type: "string" },
+            created: { type: "string" }
```
As previously mentioned I don't think that's a good idea. :)

wlach commented on src/components/modals/revision-diff.jsx:
```diff
@@ -0,0 +1,41 @@
+    ) {
+      return <p>Getting revisions...</p>;
+    } else if (
+      this.props.previousRevisionContent.localeCompare(
```
Yeah I think this is actually not necessary, now that I look at it. I'll change it.

wlach commented on src/components/modals/history-modal.jsx:
```diff
@@ -0,0 +1,155 @@
+        {this.props.gettingRevisionList && <p>Getting revisions...</p>}
+        {this.props.errorGettingRevisionList && <p>Error getting revisions</p>}
+        {!this.props.gettingRevisionList &&
+          !this.props.errorGettingRevisionList && (
```
This seems reasonable, I'll give it a shot

wlach commented on src/components/modals/history-modal.jsx:
```diff
@@ -0,0 +1,155 @@
+    gettingRevisionContent
+  } = notebookHistory;
+
+  let { currentRevisionContent, previousRevisionContent } = {};
```
Yeah this was silly. :) Fixed in my update.

wlach commented:
Modified per discussion. I think this is feature complete at this point. I'd like to file some issues for the things that @hamilton brought up and address them in followups.

wlach commented:
Oops, sorry about that! I'll add it now and then merge.

hamilton commented:
Looks good @wlach! We can file follow-up issues for further improvements & additions, but I'm going to merge this.
